### ISSUE TITLE
Interface with HIDGuardian

### ISSUE THREAD

**Issue by WhiteMagic** [WhiteMagic/JoystickGremlin#63]
HIDGuardian (https://github.com/nefarius/ViGEm/tree/master/HidGuardian) allows hiding of physical devices similar to what TARGET etc. can do but for any device. As it seems to be mainly configured via registry settings it should be quite easy to interface to it using https://docs.python.org/3.6/library/winreg.html?highlight=winreg

**Comment by Wolfereign** [comment]
Just to give some more info on this since the documentation at ViGEm seems light.  The filter driver that HIDGuardian installs is controlled primarily by registry values.  Hidden devices are controlled by a "REG_MULTI_SZ" value called "AffectedDevices".  This entry contains the hardware ids for the devices you want to hide. To add programs to the whitelist to access the hidden devices you just need to add the Proccess ID of the application as a Key to the whitelist key in the registry. 
![image](https://user-images.githubusercontent.com/5723368/34124485-4f53d68c-e3f0-11e7-8554-6df31bc77395.png)


So all Joystick Gremlin would need to do is install the HIDGuardian Driver. Then when the profile is activated update the Whitelist with Joystick Gremlins current Proccessor ID. Then add the devices that are part of the Profile to the "AffectedDevices".  When the profile is deactivated clear the devices from the "AfffectedDevices".  evilC wrote an autohotkey script that installs HID Guardian if you want a reference https://autohotkey.com/boards/viewtopic.php?t=34890.  Just to clarify HIDCerberus is not needed.  Sense we are changing setting using the registry directly HIDCerberus would be redundant.

**Comment by WhiteMagic** [comment]
Thanks for doing a bunch of the sifting through information on this. The way you describe is essentially what I had in mind, i.e. have Gremlin directly fiddle with registry entries to configure things. I hadn't looked deeper into what I needed to modify or how feasibly this would be but by the looks of it that doesn't seem too excessive.

**Comment by Wolfereign** [comment]
You will have to create the Registry Key yourself because the driver doesn't create it.  It just checks for it as far as I can see
...\HidGuardian
...\HidGuardian\Parameters
...\HidGuardian\Parameters\AffectedDevices
...\HidGuardian\Parameters\Whitelist
...\HidGuardian\Parameters\Whitelist\ProgramPID

**Comment by Wolfereign** [comment]
I wrote a python wrapper for HidGuardian.  https://gist.github.com/Wolfereign/7fdee8112cc12c184676360b12d054e8#file-hidguardian-py

Its it needs some documentation but the core of it is there.

The HidGuardian class isn't finished but the AffectedDevices and WhiteList classes to manipulate the registry are finished.

**Comment by WhiteMagic** [comment]
Thanks, I've actually played with the exact same thing yesterday and put together something to do the same type of registry mangling. I'm currently plugging it into Gremlin to add the UI side to it and figure out how and what to expose to the user.

I'll take a look at your code which seems to follow the structure of the registry entries more closely then what I wrote, which is more looking at it from a doing things point of view. Having a second implementation definitely will make double checking all the error handling easier since there's probably a few odd cases I haven't looked at.
